File: c/db/connection.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>

#include "connection.h"

Connection*
open_connection(char* filename)
{
  int fd = open(filename, O_RDWR | O_CREAT, S_IWUSR | S_IRUSR);
  if (fd == -1) {
    printf("Impossible d'ouvrir le fichier '%s'\n", filename);
    exit(EXIT_FAILURE);
  }

  Connection* connection = malloc(sizeof(Connection));
  connection->fd = fd;
  return connection;
}

File: rust/db/connection.rs
use std::fs::File;

pub(crate) struct Connection {
    fd: File,
}

impl Connection {
    pub(crate) fn open(database_filename: &str) -> Self {
        let fd = File::open(database_filename)
            .or_else(|_| File::create(database_filename))
            .expect("Erreur: impossible d'ouvrir le fichier.");

        Self { fd }
    }
}
